globalsat_sport.cc

```diff
@@ -0,0 +1,742 @@
+	if (global_opts.debug_level > 1) {
+		printf ("DeviceCommand: 0x%02x ", DeviceCommand);
+	}
+	len_h = serial_recv_byte ();
+	calc_crc ^= len_h;
+	len_l = serial_recv_byte ();
+	calc_crc ^= len_l;
+
+	lenght = (len_h << 8) + len_l;
+	if (global_opts.debug_level > 1) {
+		printf ("len=%d Payload:", lenght);
+	}
+
+	payload = (uint8_t *) malloc (lenght);
+	if (payload == NULL) {
+		goto error_out;
+	}
+
+	for (i = 0; i < lenght; i++) {
+		payload[i] = serial_recv_byte ();
+		calc_crc ^= payload[i];
+	}
+
+	crc = serial_recv_byte ();
+	if (global_opts.debug_level > 1) {
+		printf ("crc=0x%x should be=0x%x\n", crc, calc_crc);
+	}
+	if (crc == calc_crc) {
+		*out_DeviceCommand = DeviceCommand;
+		*out_lenght = lenght;
+		return payload;
+	}
+	//crc error
+	free (payload);
+	error_out:
+	*out_lenght = 0;
+	return NULL;
+}
+
+/*
+ * Send on byte package
+ */
+static void
+globalsat_send_simple (uint8_t command)
+{
+	uint8_t payload[1];
+	payload[0] = command;
+	globalsat_write_package (payload, 1);
+}
+
+static void
+globalsat_probe_device ()
+{
+	//2 1 0 191 190
+
+	//TODO try this first if fails try with false, to support 561
+	isSizeSwaped = FALSE;		//all but 561
+
+	globalsat_send_simple (CommandWhoAmI);
+
+	uint8_t *payload;
+	int len;
+	uint8_t DeviceCommand;
+	payload = globalsat_read_package (&len, &DeviceCommand);
+	if ((len > 0) && (payload != NULL)) {
+		if (global_opts.debug_level > 1) {
+			printf ("Got package!!!\n");
+		}
+		//TODO figure out what device it is if we start to support more devices then gh625XT
+	}
+
+	if (payload) {
+		free (payload);
+	}
+}
+
+static void
+rd_init (const char *fname)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " rd_init()\n");
+	}
+	serial_init (fname);
+	globalsat_probe_device ();
+}
+
+static void
+wr_init(const char *fname)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " wr_init()\n");
+	}
+	serial_init (fname);
+}
+
+
+static void
+rd_deinit(void)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " rd_deinit()\n");
+	}
+	serial_deinit();
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " rd_deinit() Done\n");
+	}
+}
+
+static void
+wr_deinit(void)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " wr_deinit()\n");
+	}
+	serial_deinit ();
+}
+
+static void track_read (void);
+
+
+static void
+waypoint_read(void)
+{
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME "   waypoint_read()\n");
+	}
+	//CommandGetTrackFileHeaders
+	globalsat_send_simple (CommandGetWaypoints);
+
+	uint8_t *in_payload;
+	int len;
+	uint8_t DeviceCommand;
+	in_payload = globalsat_read_package (&len, &DeviceCommand);
+	if ((len > 0) && (in_payload != NULL)) {
+		if (global_opts.debug_level > 1) {
+			printf ("Got package!!!\n");
+		}
+		//TODO figure out what device it is
+	}
+	if (in_payload) {
+		free (in_payload);
+	}
+	track_read ();
+}
+
+static void
+track_read(void)
+{
+	route_head *trk = NULL;
+	Waypoint *wpt = NULL;
+    gpsbabel::DateTime gpsDateTime;
```

Comment:
Nit: For readability, you can move the decls to their first use which also cleans up the default initializer case.

Some day we can use C++11 which allows 'auto'...